File: packages/domain/src/snapshot.ts
import { z } from 'zod';

import { safeForegroundFor } from './color';
import { isColorOnly, type ChromaticMemory } from './memory';
import { hexSchema } from './palette';

/**
 * What a native extension is allowed to know about a memory.
 *
 * **This file is a wire format, not a model.** Everything here crosses out of
 * JavaScript into a WidgetKit or ActivityKit process that has no React Native
 * runtime, no MMKV, no access to the app's documents directory and no way to ask
 * a question. It gets one JSON file in a shared App Group container and must
 * render something correct from whatever it finds there — including nothing.
 *
 * Three rules follow, and every decision below is a consequence of one of them:
 *
 *  1. **The extension never computes.** Contrast, formatting, ranking and
 *     truncation all happen here, where they are tested. A widget that has to
 *     decide whether white or black reads on a colour is a widget that will get
 *     it wrong on somebody's lock screen at 6am.
 *  2. **The extension never fetches.** No remote URL appears in this payload —
 *     see `artworkPath`. WidgetKit renders on a timeline the app does not
 *     control, frequently with no network, and a card that renders a hole is
 *     worse than one that renders a colour.
 *  3. **The extension never learns anything private.** Notes, titles, locations
 *     and tags are absent by construction rather than by filtering. See
 *     `toWidgetSnapshot`.
 *
 * Mirrors `Snapshot.swift`. The two are checked against each other by
 * `snapshot.test.ts`'s Codable-compatibility suite, which is the only thing
 * standing between a renamed field here and a blank widget in production.
 */

/**
 * Bumped when a *reader* would misinterpret the previous shape.
 *
 * Adding an optional field that old Swift ignores does not need a bump; renaming
 * or re-meaning one does. The reader refuses anything it does not recognise
 * rather than guessing — see `readWidgetSnapshotFile`.
 */
export const WIDGET_SNAPSHOT_VERSION = 1;

/** The app's URL scheme, from `app.json`. Deep links are built from it here so
 *  the extension only ever copies a string it was handed. */
export const DEEP_LINK_SCHEME = 'chromawave';

/**
 * How many memories the file carries.
 *
 * A large widget shows at most six, and the shared container is read
 * synchronously on the extension's main thread during a timeline request. This
 * is a rendering budget, not a library.
 */
export const WIDGET_SNAPSHOT_CAPACITY = 8;

/** Bands a widget draws. Beyond five they are thinner than a hairline at small size. */
export const WIDGET_SNAPSHOT_BANDS = 5;

/* --------------------------------------------------------------- deep links */

export const deepLinkTargets = ['memory', 'pair', 'today', 'library'] as const;
export type DeepLinkTarget = (typeof deepLinkTargets)[number];

export type DeepLink =
  | { target: 'memory'; memoryId: string }
  | { target: 'pair'; memoryId: string }
  | { target: 'today' }
  | { target: 'library' };

/**
 * A link into the app, as a string both Swift and `expo-router` accept.
 *
 * Built rather than written, because the two places that need one — a widget's
 * `widgetURL` and a Live Activity's tap target — are in a language that cannot
 * import this module, and a hand-typed `chromawave://memroy/…` fails silently.
 */
export function toDeepLink(link: DeepLink): string {
  switch (link.target) {
    case 'memory':
      return `${DEEP_LINK_SCHEME}://memory/${encodeURIComponent(link.memoryId)}`;
    case 'pair':
      return `${DEEP_LINK_SCHEME}://pair/${encodeURIComponent(link.memoryId)}`;
    case 'today':
      return `${DEEP_LINK_SCHEME}://today`;
    case 'library':
      return `${DEEP_LINK_SCHEME}://library`;
  }
}

/**
 * The inverse, and deliberately strict.
 *
 * A deep link is untrusted input: it arrives from the OS, and on iOS anything
 * that can open a URL can send one. An unrecognised host, a wrong scheme or a
 * missing id returns null and the app stays where it is, which is the same
 * allowlist posture `useNotificationRoute` already takes with payload routes.
 */
export function parseDeepLink(url: string): DeepLink | null {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return null;
  }

  if (parsed.protocol !== `${DEEP_LINK_SCHEME}:`) return null;

  // `chromawave://memory/abc` parses with host 'memory' and pathname '/abc'.
  const host = parsed.hostname;
  const id = decodeURIComponent(parsed.pathname.replace(/^\/+/, '')).trim();

  switch (host) {
    case 'memory':
      return id ? { target: 'memory', memoryId: id } : null;
    case 'pair':
      return id ? { target: 'pair', memoryId: id } : null;
    // A trailing segment on a target that takes no id is a malformed link, not
    // a link with something extra — accepting it would make two different URLs
    // mean the same thing and a typo indistinguishable from intent.
    case 'today':
      return id ? null : { target: 'today' };
    case 'library':
      return id ? null : { target: 'library' };
    default:
      return null;
  }
}

/* ----------------------------------------------------------------- snapshot */

export const widgetBandSchema = z.object({
  hex: hexSchema,
  /** 0–1, already normalised. Swift lays these out with no arithmetic. */
  weight: z.number().min(0).max(1),
});

export type WidgetBand = z.infer<typeof widgetBandSchema>;

export const widgetMemorySnapshotSchema = z.object({
  id: z.string().uuid(),

  /**
   * Group-container-relative, never absolute.
   *
   * The extension resolves it against its own
   * `containerURL(forSecurityApplicationGroupIdentifier:)`, because the app's
   * absolute paths contain a data-container UUID that iOS is free to change
   * between launches — an absolute path baked into a snapshot is a broken image
   * one OS update later. Null for a colour-only memory, which renders as bands.
   */
  imagePath: z.string().min(1).max(200).nullable(),

  /**
   * Also group-relative, and null until the app has cached the artwork itself.
   *
   * The track's remote `artworkUrl` is deliberately *not* carried. Rule 2: a
   * widget must not reach the network to draw, and the provider's terms do not
   * contemplate an extension hot-linking their CDN on a timeline refresh.
   */
  artworkPath: z.string().min(1).max(200).nullable(),

  bands: z.array(widgetBandSchema).min(1).max(WIDGET_SNAPSHOT_BANDS),

  /** The band a solid fill should use. Precomputed so Swift never picks. */
  dominantHex: hexSchema,
  /** Black or white, already contrast-checked against `dominantHex`. Rule 1. */
  inkHex: hexSchema,

  /** From `AtmosphereReading.mood`. A short lowercase word; Swift capitalises. */
  mood: z.string().min(1).max(24),

  /** Null when the memory is unpaired. Both are null together, never one. */
  trackTitle: z.string().min(1).max(80).nullable(),
  trackArtist: z.string().min(1).max(80).nullable(),

  /** ISO 8601. Formatted on the device, in the device's locale, by Swift. */
  capturedAt: z.iso.datetime(),

  /** Where a tap goes. Always populated — a widget with no destination is a poster. */
  deepLink: z.string().min(1).max(300),
});

export type WidgetMemorySnapshot = z.infer<typeof widgetMemorySnapshotSchema>;

export const widgetSnapshotFileSchema = z.object({
  schemaVersion: z.literal(WIDGET_SNAPSHOT_VERSION),
  /** When the app last wrote this. The extension shows staleness from it. */
  updatedAt: z.iso.datetime(),
  /** So a widget can match the app the user actually looks at. */
  skin: z.enum(['chroma', 'swiss']),
  entries: z.array(widgetMemorySnapshotSchema).max(WIDGET_SNAPSHOT_CAPACITY),
});

export type WidgetSnapshotFile = z.infer<typeof widgetSnapshotFileSchema>;

/**
 * The empty file, which is a real state and not an error.
 *
 * A fresh install has a widget on the home screen before it has a memory in the
 * library. That widget renders the empty case, and it does so from this rather
 * than from a nil the extension has to branch on in four places.
 */
export const emptyWidgetSnapshotFile = (
  skin: WidgetSnapshotFile['skin'] = 'chroma',
  now: string = new Date().toISOString(),
): WidgetSnapshotFile => ({
  schemaVersion: WIDGET_SNAPSHOT_VERSION,
  updatedAt: now,
  skin,
  entries: [],
});

/* --------------------------------------------------------------- projection */

/**
 * One memory, reduced to what a lock screen may show.
 *
 * **What is dropped, and why it is dropped here rather than filtered later.**
 * `personalContext` — the note, the title, the mood the user typed, the tags,
 * the place name — never enters this function's output. Not because a widget
 * would look cluttered, but because a widget is visible to anyone who picks up
 * the phone, and the note is the one field in `ChromaticMemory` that contains
 * whatever the user felt like writing. A `redacted` flag would be the wrong
 * mechanism: it can be forgotten, inverted, or read after the data has already
 * been written to a container that other processes can open. Absence cannot.
 *
 * `visualAnalysis`, `facets`, `collectionIds` and the recommendation list are
 * dropped for size: none of them is drawable at 155pt.
 */
export function toWidgetSnapshot(memory: ChromaticMemory): WidgetMemorySnapshot {
  const bands = [...memory.palette.colors]
    // Widest first, so a truncated set keeps the colours the photograph was
    // actually made of rather than the first five the extractor happened to emit.
    .sort((first, second) => second.weight - first.weight)
    .slice(0, WIDGET_SNAPSHOT_BANDS);

  // Re-normalised after truncation. Swift lays bands out proportionally and a
  // set summing to 0.82 would leave a gap that looks like a rendering bug.
  const total = bands.reduce((sum, color) => sum + color.weight, 0) || 1;

  const dominantHex =
    memory.palette.colors.find((color) => color.role === 'dominant')?.hex ??
    bands[0]?.hex ??
    '#000000';

  const track = memory.musicPairing.selectedTrack;

  return {
    id: memory.id,
    // A colour-only memory has no frame to show; `isColorOnly` is the same
    // predicate the library uses, so the two never disagree about what a
    // palette-only record looks like.
    imagePath: isColorOnly(memory.image) ? null : imagePathFor(memory.id),
    // Populated by the snapshot writer once artwork has been cached into the
    // group container. Null here is correct and common: most memories are read
    // before their artwork has ever been fetched.
    artworkPath: null,
    bands: bands.map((color) => ({
      hex: color.hex,
      weight: Math.round((color.weight / total) * 1000) / 1000,
    })),
    dominantHex,
    inkHex: safeForegroundFor(dominantHex),
    mood: memory.atmosphere.mood,
    // Both or neither. The schema allows each to be null independently; this is
    // the only writer, and it keeps them in step so Swift can test one field.
    trackTitle: track ? truncate(track.title, 80) : null,
    trackArtist: track ? truncate(track.artist, 80) : null,
    capturedAt: memory.capturedAt,
    deepLink: toDeepLink({ target: 'memory', memoryId: memory.id }),
  };
}

/**
 * The group-relative path the app writes a widget's copy of a frame to.
 *
 * A *copy*, resized for the largest widget family, not the original: the
 * original is a multi-megabyte camera frame in the app's documents directory,
 * which an extension cannot read and should not be asked to decode on a
 * timeline request.
 */
export const imagePathFor = (memoryId: string): string => `widget/${memoryId}.jpg`;
export const artworkPathFor = (memoryId: string): string => `widget/${memoryId}-artwork.jpg`;

/**
 * The file the app writes, from the memories it has.
 *
 * Ordering is the product decision here: most recently captured first, because
 * the small widget shows `entries[0]` and "the last thing I saw" is the only
 * ordering a person can predict without opening the app.
 */
export function buildWidgetSnapshotFile(
  memories: readonly ChromaticMemory[],
  options: { skin: WidgetSnapshotFile['skin']; now?: string },
): WidgetSnapshotFile {
  const entries = [...memories]
    .sort((first, second) => second.capturedAt.localeCompare(first.capturedAt))
    .slice(0, WIDGET_SNAPSHOT_CAPACITY)
    .map(toWidgetSnapshot);

  return {
    schemaVersion: WIDGET_SNAPSHOT_VERSION,
    updatedAt: options.now ?? new Date().toISOString(),
    skin: options.skin,
    entries,
  };
}

/* ------------------------------------------------------------------ reading */

export type SnapshotReadResult =
  | { status: 'ok'; file: WidgetSnapshotFile; dropped: number }
  | { status: 'empty' }
  /** A version this build does not understand. Refused, never guessed at. */
  | { status: 'unsupported'; found: number }
  | { status: 'corrupt'; reason: string };

/**
 * Parses a snapshot file defensively, the way the extension has to.
 *
 * Exists in TypeScript as well as Swift because it is the behaviour worth
 * testing: the Swift side is thirty lines of `JSONDecoder` and the interesting
 * cases — a half-written file, a future version, one bad entry among seven — are
 * all decided by this logic. `snapshot.test.ts` exercises them here so the Swift
 * port has a specification rather than an intention.
 *
 * A single unreadable entry is dropped and the rest are kept, which is the same
 * per-record posture `StoredMemoryRepository` takes: losing one memory from a
 * widget is a blemish, losing all eight is a bug report.
 */
export function readWidgetSnapshotFile(raw: unknown): SnapshotReadResult {
  if (raw === null || raw === undefined || raw === '') return { status: 'empty' };

  let value: unknown = raw;
  if (typeof raw === 'string') {
    try {
      value = JSON.parse(raw);
    } catch {
      // A truncated write. The app writes atomically, so this means the file was
      // damaged outside our control — recoverable by the next write.
      return { status: 'corrupt', reason: 'not-json' };
    }
  }

  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    return { status: 'corrupt', reason: 'not-an-object' };
  }

  const version = (value as { schemaVersion?: unknown }).schemaVersion;
  if (typeof version !== 'number' || !Number.isInteger(version)) {
    return { status: 'corrupt', reason: 'no-version' };
  }
  // Older *and* newer are both refused. An extension is updated with the app, so
  // a mismatch means a downgrade or a partially-installed update, and rendering
  // a best-effort guess at an unknown shape is how wrong data reaches a lock
  // screen. Refusing shows the empty state, which is honest.
  if (version !== WIDGET_SNAPSHOT_VERSION) return { status: 'unsupported', found: version };

  const envelope = z
    .object({
      schemaVersion: z.literal(WIDGET_SNAPSHOT_VERSION),
      updatedAt: z.iso.datetime(),
      skin: z.enum(['chroma', 'swiss']),
      entries: z.array(z.unknown()),
    })
    .safeParse(value);

  if (!envelope.success) {
    return { status: 'corrupt', reason: 'envelope-invalid' };
  }

  const entries: WidgetMemorySnapshot[] = [];
  let dropped = 0;
  for (const candidate of envelope.data.entries.slice(0, WIDGET_SNAPSHOT_CAPACITY)) {
    const parsed = widgetMemorySnapshotSchema.safeParse(candidate);
    if (parsed.success) entries.push(parsed.data);
    else dropped++;
  }

  return {
    status: 'ok',
    file: { ...envelope.data, entries },
    dropped,
  };
}

/**
 * Whether the app should ask WidgetKit to reload.
 *
 * WidgetKit budgets timeline reloads per day and silently starts ignoring an app
 * that asks too often — so "reload after every write" is not a safe default, it
 * is how a widget stops updating for everyone. Content-compared rather than
 * timestamp-compared, because `updatedAt` changes on every write by definition
 * and would make this function always return true.
 */
export function snapshotChanged(
  previous: WidgetSnapshotFile | null,
  next: WidgetSnapshotFile,
): boolean {
  if (previous === null) return true;
  if (previous.skin !== next.skin) return true;
  if (previous.entries.length !== next.entries.length) return true;
  return previous.entries.some((entry, index) => !sameEntry(entry, next.entries[index]));
}

const sameEntry = (first: WidgetMemorySnapshot, second: WidgetMemorySnapshot | undefined) =>
  second !== undefined &&
  first.id === second.id &&
  first.imagePath === second.imagePath &&
  first.artworkPath === second.artworkPath &&
  first.dominantHex === second.dominantHex &&
  first.mood === second.mood &&
  first.trackTitle === second.trackTitle &&
  first.trackArtist === second.trackArtist &&
  first.bands.length === second.bands.length &&
  first.bands.every(
    (band, index) =>
      band.hex === second.bands[index]?.hex && band.weight === second.bands[index]?.weight,
  );

/** Cuts on a word boundary where there is one nearby, so a title ends in a word. */
function truncate(value: string, max: number): string {
  const trimmed = value.trim();
  if (trimmed.length <= max) return trimmed;
  const cut = trimmed.slice(0, max - 1);
  const space = cut.lastIndexOf(' ');
  return `${space > max * 0.6 ? cut.slice(0, space) : cut}…`;
}

File: apps/mobile/targets/ChromawaveWidget/Snapshot.swift
import Foundation
import SwiftUI
import UIKit

/// The Swift half of `packages/domain/src/snapshot.ts`.
///
/// Every field here has a counterpart there and the two are checked against each
/// other by that file's "Codable compatibility" suite. Synthesised `Decodable`
/// fails the *whole* decode on a key it expects and cannot find, so the TS side
/// emits every key explicitly — including nulls — and this side declares the
/// nullable ones as optionals. Neither half may change alone.
///
/// Nothing in this file computes anything a widget could get wrong: contrast,
/// truncation, ordering and band weights all arrive already decided.

let kSchemaVersion = 1
let kSnapshotFileName = "widget-snapshot.json"

struct WidgetBand: Codable, Hashable {
  let hex: String
  /// 0–1, already normalised to sum to one across the set.
  let weight: Double
}

struct WidgetMemorySnapshot: Codable, Hashable, Identifiable {
  let id: String
  /// Relative to the App Group container. Never absolute — the app's data
  /// container UUID changes between installs.
  let imagePath: String?
  let artworkPath: String?
  let bands: [WidgetBand]
  let dominantHex: String
  /// Already contrast-checked against `dominantHex` on the JavaScript side.
  let inkHex: String
  let mood: String
  let trackTitle: String?
  let trackArtist: String?
  /// ISO 8601. Formatted here, in the viewer's locale.
  let capturedAt: String
  let deepLink: String
}

struct WidgetSnapshotFile: Codable {
  let schemaVersion: Int
  let updatedAt: String
  let skin: String
  let entries: [WidgetMemorySnapshot]
}

// MARK: - Reading

/// Mirrors `readWidgetSnapshotFile`, including its refusals.
///
/// The important behaviour is what it does *not* do: an unrecognised
/// `schemaVersion` is refused rather than best-effort decoded, because a widget
/// that renders a confident guess at an unknown shape puts wrong information on
/// somebody's lock screen. Refusing shows the empty state, which is honest.
enum SnapshotRead {
  case ok(WidgetSnapshotFile)
  case empty
  case unsupported(Int)
  case corrupt

  var file: WidgetSnapshotFile? {
    if case let .ok(file) = self { return file }
    return nil
  }
}

enum SnapshotStore {
  /// Resolved from the extension's own entitlement rather than passed in, so the
  /// widget and the app cannot disagree about which container they mean.
  static var appGroup: String {
    guard
      let identifier = Bundle.main.object(forInfoDictionaryKey: "CFBundleIdentifier") as? String
    else { return "group.com.chromawave.app" }
    // `com.chromawave.app.widget` → `group.com.chromawave.app`
    let base = identifier.hasSuffix(".widget") ? String(identifier.dropLast(7)) : identifier
    return "group.\(base)"
  }

  static var containerURL: URL? {
    FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroup)
  }

  static func read() -> SnapshotRead {
    guard let container = containerURL else { return .empty }
    let url = container.appendingPathComponent(kSnapshotFileName)

    guard let data = try? Data(contentsOf: url), !data.isEmpty else { return .empty }

    // The version is read before the body, so a future file is refused rather
    // than producing a decode error that looks like corruption.
    guard
      let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let version = object["schemaVersion"] as? Int
    else { return .corrupt }

    guard version == kSchemaVersion else { return .unsupported(version) }

    guard let file = try? JSONDecoder().decode(WidgetSnapshotFile.self, from: data) else {
      return .corrupt
    }
    return .ok(file)
  }

  /// An image the app cached into the shared container for exactly this purpose.
  ///
  /// Returns nil rather than reaching for the original: the app's copy is
  /// already resized for the largest family, and decoding a camera frame during
  /// a timeline request is how a widget gets killed for exceeding its memory
  /// limit.
  static func image(at relativePath: String?) -> UIImage? {
    guard let relativePath, let container = containerURL else { return nil }
    let url = container.appendingPathComponent(relativePath)
    guard let data = try? Data(contentsOf: url) else { return nil }
    return UIImage(data: data)
  }
}

// MARK: - Presentation helpers

extension String {
  /// `#7C5CFF` → Color. Returns a mid grey for anything unparseable, because a
  /// widget must draw something and a crash in a timeline provider is a blank
  /// tile the user cannot dismiss.
  var chromaColor: Color {
    var hex = trimmingCharacters(in: .whitespacesAndNewlines)
    if hex.hasPrefix("#") { hex.removeFirst() }
    guard hex.count == 6, let value = UInt64(hex, radix: 16) else {
      return Color(white: 0.5)
    }
    return Color(
      .sRGB,
      red: Double((value & 0xFF0000) >> 16) / 255,
      green: Double((value & 0x00FF00) >> 8) / 255,
      blue: Double(value & 0x0000FF) / 255,
      opacity: 1
    )
  }
}

extension WidgetMemorySnapshot {
  var capturedDate: Date? {
    ISO8601DateFormatter.chromawave.date(from: capturedAt)
  }

  /// "3 days ago" in the viewer's language, or a plain date when it is old
  /// enough that a relative phrase stops being useful.
  var displayDate: String {
    guard let date = capturedDate else { return "" }
    let elapsed = Date().timeIntervalSince(date)
    if elapsed < 60 * 60 * 24 * 7 {
      let formatter = RelativeDateTimeFormatter()
      formatter.unitsStyle = .abbreviated
      return formatter.localizedString(for: date, relativeTo: Date())
    }
    return date.formatted(.dateTime.day().month(.abbreviated))
  }

  /// What the widget calls the memory when there is no track yet. Never the
  /// user's own note — that never crosses into this process. See `snapshot.ts`.
  var headline: String { trackTitle ?? mood.capitalized }
  var subhead: String? { trackArtist }
  var isPaired: Bool { trackTitle != nil }

  var url: URL? { URL(string: deepLink) }
}

extension ISO8601DateFormatter {
  /// The app writes fractional seconds; the default formatter rejects them, and
  /// a nil date here silently blanks every timestamp in every widget family.
  static let chromawave: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter
  }()
}

// MARK: - Fixtures

/// Placeholder and preview content.
///
/// Deliberately real-looking rather than lorem: a placeholder with a one-word
/// title hides exactly the layout bug — a long title colliding with the date —
/// that shipping without one guarantees.
extension WidgetMemorySnapshot {
  static let placeholder = WidgetMemorySnapshot(
    id: "00000000-0000-4000-8000-000000000000",
    imagePath: nil,
    artworkPath: nil,
    bands: [
      WidgetBand(hex: "#2A2352", weight: 0.42),
      WidgetBand(hex: "#7C5CFF", weight: 0.26),
      WidgetBand(hex: "#22D3EE", weight: 0.17),
      WidgetBand(hex: "#F4A38C", weight: 0.15),
    ],
    dominantHex: "#2A2352",
    inkHex: "#FFFFFF",
    mood: "nocturnal",
    trackTitle: "Nightswimming",
    trackArtist: "R.E.M.",
    capturedAt: ISO8601DateFormatter.chromawave.string(from: Date()),
    deepLink: "chromawave://memory/00000000-0000-4000-8000-000000000000"
  )

  /// The unpaired case, which is the one most likely to be laid out badly
  /// because it is the one nobody screenshots.
  static let unpaired = WidgetMemorySnapshot(
    id: "00000000-0000-4000-8000-000000000001",
    imagePath: nil,
    artworkPath: nil,
    bands: [
      WidgetBand(hex: "#F2F1EE", weight: 0.55),
      WidgetBand(hex: "#101010", weight: 0.30),
      WidgetBand(hex: "#D8352A", weight: 0.15),
    ],
    dominantHex: "#F2F1EE",
    inkHex: "#09090B",
    mood: "austere",
    trackTitle: nil,
    trackArtist: nil,
    capturedAt: ISO8601DateFormatter.chromawave.string(from: Date()),
    deepLink: "chromawave://memory/00000000-0000-4000-8000-000000000001"
  )
}
